dockerworker/artifact.go

```diff
@@ -0,0 +1,151 @@
+// +build docker
+
+package dockerworker
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/mattetti/filebuffer"
+	"github.com/mitchellh/ioprogress"
+	"github.com/taskcluster/taskcluster-client-go/tcqueue"
+	artifact "github.com/taskcluster/taskcluster-lib-artifact-go"
+)
+
+// ScheduleReclaim run a goroutine that will perform reclaims in the background.
+func (d *DockerWorker) ScheduleReclaim(claim *tcqueue.TaskClaimResponse) {
+	takenUntil := claim.TakenUntil
+
+	go func() {
+		for {
+			select {
+			case <-time.After(time.Time(takenUntil).Sub(time.Now())):
+				reclaim, err := d.queue.ReclaimTask(claim.Status.TaskID, fmt.Sprint(claim.RunID))
+				if err != nil {
+					d.TaskLogger.Printf("Exiting reclaim loop: %v", err)
+					return
+				}
+				takenUntil = reclaim.TakenUntil
+			case <-d.Context.Done():
+				return
+			}
+		}
+	}()
+}
+
+// DownloadArtifact downloads an artifact using exponential backoff algorithm
+func (d *DockerWorker) DownloadArtifact(taskID, runID, name string, out io.WriteSeeker) (ret error) {
```

Comment:
Why is this logic here too? It also exists already, and is unrelated to docker engine changes.

Reply:
It is only used by docker to download artifact images and for tests.

Reply:
OK, please remove it, it isn't needed.

Reply:
`DownloadArtifact` is used to download docker images. If you are concerned because it is part of the package public interface, I can make it an internal function.